File: lib/aura/cli/shell/console_renderer.rb
# frozen_string_literal: true

module Aura
  module CLI
    module Shell
      class ConsoleRenderer
        def initialize(verbose: false)
          @verbose = verbose
          @last_streamed = false
        end

        def on_token(text)
          unless @last_streamed
            @last_streamed = true
            print "\r\e[K" # Clear waiting line
          end
          print text
          $stdout.flush
        end

        def on_stream_end
          puts "" if @last_streamed
          @last_streamed = false
        end

        def on_waiting(elapsed)
          print "\r⏳ Waiting for response... (#{format("%.1fs", elapsed)})"
          $stdout.flush
        end

        def on_clear_waiting
          print "\r\e[K"
        end

        def on_tool_start(tool, summary, args)
          puts "\n>> 🔧 Tool: #{tool}"
          puts "   🧾 Summary: #{summary}" if summary && !summary.to_s.strip.empty?
          if @verbose
            puts "   🧩 Args: #{format_args(args)}" unless args.nil? || args.empty?
          end
        end

        def on_tool_executing
          puts "   🚀 Executing..."
        end

        def on_tool_result(result)
          status = result.is_a?(Hash) ? (result["status"] || result[:status]) : nil
          status_color = case status.to_s
          when "ok", "success" then "\e[32m"  # Green
          when "failed", "blocked" then "\e[31m"  # Red
          else "\e[33m"  # Yellow
          end
          
          puts "   #{status_color}✓ Status: #{status}\e[0m"
          
          # Extract and display output
          output = nil
          if result.is_a?(Hash)
            output = result["output"] || result[:output] ||
                     result["content"] || result[:content] ||
                     result["stdout"] || result[:stdout] ||
                     result["message"] || result[:message]
          end
          
          if output && !output.to_s.strip.empty?
            output_str = output.to_s.strip
            # Truncate long outputs
            if output_str.length > 200
              output_str = output_str[0..197] + "..."
            end
            # Show first line or truncated output
            first_line = output_str.lines.first&.strip || output_str
            puts "   📄 #{first_line}" if first_line && !first_line.empty?
          end
          
          # Display modified files if present
          if result.is_a?(Hash)
            modified = result["modified_files"] || result[:modified_files]
            if modified && !modified.empty?
              puts "   📝 Modified files:"
              modified.each do |file|
                puts "      • #{file}"
              end
            end
          end
        end

        def on_thought(thought, elapsed = nil)
          if elapsed
            puts "\n>> 💬 Response (#{format_duration(elapsed)}):"
          else
            puts "\n>> 💬 Response:"
          end
          puts thought.to_s
        end

        def on_error(message)
          puts "\n>> ⚠️  Error: #{message}"
        end

        def on_warning(message)
           puts "\n>> ⚠️  #{message}"
        end

        def ask_confirmation(message)
          print "   ⚠️  #{message} [y/N] "
          answer = $stdin.gets.strip.downcase
          answer == "y"
        end

        private

        def format_duration(seconds)
          return "0s" unless seconds
          if seconds < 60
            format("%.1fs", seconds)
          else
            format("%dm %ds", seconds / 60, seconds % 60)
          end
        end

        def format_args(args)
          return "" if args.nil?
          json = JSON.generate(args)
          if json.length > 100
            json[0..97] + "..."
          else
            json
          end
        rescue
          args.to_s
        end
      end
    end
  end
end

File: src/cli/shell/consoleRenderer.ts
import readline from 'node:readline';
import picocolors from 'picocolors';

export class ConsoleRenderer {
  private verbose: boolean;
  private lastStreamed = false;

  constructor(options: { verbose?: boolean } = {}) {
    this.verbose = options.verbose || false;
  }

  public onToken(text: string): void {
    if (!this.lastStreamed) {
      this.lastStreamed = true;
      process.stdout.write('\r\x1b[K'); // Clear waiting line
    }
    process.stdout.write(text);
  }

  public onStreamEnd(): void {
    if (this.lastStreamed) {
      process.stdout.write('\n');
    }
    this.lastStreamed = false;
  }

  public onWaiting(elapsedSeconds: number): void {
    process.stdout.write(`\r⏳ Waiting for response... (${elapsedSeconds.toFixed(1)}s)`);
  }

  public onClearWaiting(): void {
    process.stdout.write('\r\x1b[K');
  }

  public onToolStart(tool: string, summary?: string | null, args?: any): void {
    console.log(`\n>> 🔧 Tool: ${picocolors.cyan(tool)}`);
    if (summary && summary.trim()) {
      console.log(`   🧾 Summary: ${summary.trim()}`);
    }
    if (this.verbose && args && Object.keys(args).length > 0) {
      console.log(`   🧩 Args: ${this.formatArgs(args)}`);
    }
  }

  public onToolExecuting(): void {
    console.log('   🚀 Executing...');
  }

  public onToolResult(result: any): void {
    const status = result?.status;
    let statusColor = picocolors.yellow;
    if (status === 'ok' || status === 'success') {
      statusColor = picocolors.green;
    } else if (status === 'failed' || status === 'blocked') {
      statusColor = picocolors.red;
    }

    console.log(`   ${statusColor(`✓ Status: ${status}`)}`);

    const output = result?.output ?? result?.content ?? result?.stdout ?? result?.message;
    if (output && output.toString().trim()) {
      let outputStr = output.toString().trim();
      if (outputStr.length > 200) {
        outputStr = outputStr.substring(0, 197) + '...';
      }
      const firstLine = outputStr.split('\n')[0]?.trim() || outputStr;
      if (firstLine) {
        console.log(`   📄 ${firstLine}`);
      }
    }

    const modified = result?.modified_files;
    if (Array.isArray(modified) && modified.length > 0) {
      console.log('   📝 Modified files:');
      for (const file of modified) {
        console.log(`      • ${file}`);
      }
    }
  }

  public onThought(thought: string, elapsed?: number | null): void {
    if (elapsed !== undefined && elapsed !== null) {
      console.log(`\n>> 💬 Response (${this.formatDuration(elapsed)}):`);
    } else {
      console.log('\n>> 💬 Response:');
    }
    console.log(thought);
  }

  public onError(message: string): void {
    console.error(`\n>> ${picocolors.red(`⚠️  Error: ${message}`)}`);
  }

  public onWarning(message: string): void {
    console.warn(`\n>> ${picocolors.yellow(`⚠️  ${message}`)}`);
  }

  public askConfirmation(message: string): Promise<boolean> {
    return new Promise((resolve) => {
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
      });
      rl.question(`   ⚠️  ${message} [y/N] `, (answer) => {
        rl.close();
        resolve(answer.trim().toLowerCase() === 'y');
      });
    });
  }

  private formatDuration(seconds: number): string {
    if (seconds < 60) {
      return `${seconds.toFixed(1)}s`;
    }
    const m = Math.floor(seconds / 60);
    const s = Math.floor(seconds % 60);
    return `${m}m ${s}s`;
  }

  private formatArgs(args: any): string {
    try {
      const json = JSON.stringify(args);
      if (json.length > 100) {
        return json.substring(0, 97) + '...';
      }
      return json;
    } catch {
      return String(args);
    }
  }
}
